server/internal/usecase/gateway/queue.go

```diff
@@ -0,0 +1,16 @@
+package gateway
+
+import "context"
+
+//TODO: after asset has been implemented move this under asset gateway and embed Queue to Asset struct
+// *********************************
+type CompressFileExtention string
+
+const (
+	Zip = CompressFileExtention("zip")
+)
+
+// *********************************
+type Queue interface {
+	AddTask(ctx context.Context) (queueId string, err error)
+}
```

Comment:
There are several points to consider.

- It is not good if the task is not described in any detail at all at gateway. Whether it is GCP or AWS, the request to unpack a Zip itself is not the concern of the infrastructure layer, it is unquestionably the domain. The infrastructure layer is interested in how to perform the task it is handed, not what it is. Therefore, the details of the task should be describable at the domain or use case layer.
- In the description of the decompression task, it is not necessary to specify the decompression format explicitly. The actual filename of the saved asset will tell you, and if it is in a format that cannot be decompressed, it will return an error.
- You cannot get the results of a task directly. If you want the results, you will receive them indirectly by having an outside party notify the API exposed by the CMS.
- The word queue has been eliminated from the interface name. Whether or not to queue a task is a technical concern. It is even possible to execute tasks without queuing. For example, a task runner implementation could be used for unit testing, which is completed synchronously.

```go
// domain model (pkg/task)

type Payload struct {
	ExtractAsset  *ExtractAssetPayload
	CompressAsset *CompressAssetPayload
}

type ExtractAssetPayload struct {
	Asset id.AssetID
}

func (t *ExtractAssetPayload) Payload() Payload {
	return Payload {
		ExtractAsset: t
	}
}

type CompressAssetPayload struct {
	Asset id.AssetID
}

func (t *CompressAssetPayload) Payload() Payload {
	return Payload {
		CompressAsset: t
	}
}

type CompressZipPayload struct {
	Asset id.AssetID
}

type ID = id.TaskID

// gateway

type TaskRunner interface {
	Run(context.Context, Payload) (id.TaskID, error)
}
```

Reply:
I found that it is not a technical issue whether it is asynchronous or not, but that the process of decompression is placed as business logic. We have reflected the review and ask that you review it again. 😄 

I fixed where you pointed out with following commit [6ae4627](https://github.com/reearth/reearth-cms/pull/109/commits/6ae46276c4c8105f5906c301a3e53e96b0172c12)